subsys/bluetooth/services/ras/rreq/ras_rreq.c

```diff
@@ -0,0 +1,714 @@
+{
+	struct bt_ras_rreq *rreq = NULL;
+
+	if (ras_rreq_find(conn) != NULL) {
+		return -EALREADY;
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(m_rreq_pool); i++) {
+		if (m_rreq_pool[i].conn == NULL) {
+			rreq = &m_rreq_pool[i];
+			break;
+		}
+	}
+
+	if (rreq == NULL) {
+		return -ENOMEM;
+	}
+
+	LOG_DBG("conn %p new rreq %p", (void *)conn, (void *)rreq);
+
+	memset(rreq, 0, sizeof(struct bt_ras_rreq));
+	rreq->conn = bt_conn_ref(conn);
+
+	return 0;
+}
+
+static uint8_t ranging_data_ready_notify_func(struct bt_conn *conn,
+					      struct bt_gatt_subscribe_params *params,
+					      const void *data, uint16_t length)
+{
+
+	if (length != sizeof(uint16_t)) {
+		LOG_DBG("Ranging Data Ready Indication size error");
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	struct bt_ras_rreq *rreq = ras_rreq_find(conn);
+
+	uint16_t ranging_counter = *(uint16_t *)data;
+
+	rreq->rd_ready.cb(conn, ranging_counter);
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static void data_receive_finished(struct bt_ras_rreq *rreq)
+{
+	int error_code;
+
+	if (!rreq->on_demand_rd.last_segment_received) {
+		LOG_DBG("Ranging data completed with missing segments");
+		rreq->on_demand_rd.error_with_data_receive = true;
+	}
+
+	error_code = rreq->on_demand_rd.error_with_data_receive ? -EINVAL : 0;
+	rreq->on_demand_rd.data_get_in_progress = false;
+
+	rreq->on_demand_rd.cb(rreq->conn, rreq->on_demand_rd.counter_in_progress, error_code);
+}
+
+static uint8_t ranging_data_overwritten_notify_func(struct bt_conn *conn,
+						    struct bt_gatt_subscribe_params *params,
+						    const void *data, uint16_t length)
+{
+
+	if (length != sizeof(uint16_t)) {
+		LOG_DBG("Ranging Data Ready Indication size error");
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	struct bt_ras_rreq *rreq = ras_rreq_find(conn);
+
+	uint16_t ranging_counter = *(uint16_t *)data;
+
+	if (rreq->on_demand_rd.data_get_in_progress &&
+	    rreq->on_demand_rd.counter_in_progress == ranging_counter) {
+		if (rreq->cp.state != BT_RAS_RREQ_CP_STATE_NONE) {
+			LOG_DBG("Overwritten received while writing to RAS-CP, will continue "
+				"waiting for RAS-CP response");
+			return BT_GATT_ITER_CONTINUE;
+		}
+
+		LOG_DBG("Ranging counter %d overwritten whilst receiving", ranging_counter);
+		rreq->on_demand_rd.error_with_data_receive = true;
+		data_receive_finished(rreq);
+
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	rreq->rd_overwritten.cb(conn, ranging_counter);
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static int ack_ranging_data(struct bt_ras_rreq *rreq)
+{
+	int err;
+
+	NET_BUF_SIMPLE_DEFINE(ack_ranging_data, RASCP_CMD_OPCODE_LEN + sizeof(uint16_t));
+	net_buf_simple_add_u8(&ack_ranging_data, RASCP_OPCODE_ACK_RD);
+	net_buf_simple_add_le16(&ack_ranging_data, rreq->on_demand_rd.counter_in_progress);
+
+	err = bt_gatt_write_without_response(rreq->conn, rreq->cp.subscribe_params.value_handle,
+					     ack_ranging_data.data, ack_ranging_data.len, false);
+	if (err) {
+		LOG_DBG("CP ACK ranging data written failed, err %d", err);
```

Comment:
Would be good if a user can see this message without having to enable debug logs. Either WRN or ERR would be better imo

Reply:
hmm yeah I need to think about this - because if this fails things will go wrong. `bt_gatt_write_without_response` can fail if the queue is full when called from RX thread (which this is). So may need to move to a work queue 😞

Reply:
Alright so it's not really recoverable at the moment? If so feel free to leave it as is and defer to a TODO.

Reply:
Have ended the data get procedure here for now, peer will just have to wait to timeout unfortunately. Can think of moving to a work queue at a later date